compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AnnotationLowering.kt

```diff
@@ -6,167 +6,35 @@
 package org.jetbrains.kotlin.backend.jvm.lower
 
 import org.jetbrains.kotlin.backend.common.FileLoweringPass
-import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
 import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
 import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
-import org.jetbrains.kotlin.backend.jvm.intrinsics.IrIntrinsicMethods
-import org.jetbrains.kotlin.backend.jvm.intrinsics.KClassJavaProperty
 import org.jetbrains.kotlin.ir.IrStatement
-import org.jetbrains.kotlin.ir.builders.declarations.buildField
-import org.jetbrains.kotlin.ir.builders.irCall
-import org.jetbrains.kotlin.ir.builders.irGet
 import org.jetbrains.kotlin.ir.declarations.IrClass
 import org.jetbrains.kotlin.ir.declarations.IrConstructor
 import org.jetbrains.kotlin.ir.declarations.IrFile
-import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
-import org.jetbrains.kotlin.ir.expressions.IrCall
-import org.jetbrains.kotlin.ir.expressions.IrExpression
-import org.jetbrains.kotlin.ir.expressions.impl.IrGetterCallImpl
-import org.jetbrains.kotlin.ir.symbols.IrClassSymbol
-import org.jetbrains.kotlin.ir.symbols.IrSimpleFunctionSymbol
-import org.jetbrains.kotlin.ir.types.*
-import org.jetbrains.kotlin.ir.types.impl.makeTypeProjection
-import org.jetbrains.kotlin.ir.util.functions
 import org.jetbrains.kotlin.ir.util.isAnnotationClass
-import org.jetbrains.kotlin.ir.util.isKClass
-import org.jetbrains.kotlin.ir.util.isNonPrimitiveArray
 import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
 import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
-import org.jetbrains.kotlin.types.Variance
 
-internal val annotationPhase = makeIrFilePhase(
-    ::AnnotationLowering,
+internal val annotationPhase = makeIrFilePhase<JvmBackendContext>(
+    { AnnotationLowering() },
     name = "Annotation",
-    description = "Remove constructors and modify field types in annotation classes"
+    description = "Remove constructors of annotation classes"
 )
 
 /**
- * Remove the constructors from annotation classes and change the types of KClass
- * and Array<KClass> fields to use java.lang.Class instead. This phase also rewrites
- * the uses of annotation class fields appropriately.
+ * Remove the constructors from annotation classes.
  */
-private class AnnotationLowering(private val context: JvmBackendContext) : FileLoweringPass, IrElementTransformerVoid() {
+private class AnnotationLowering : FileLoweringPass, IrElementTransformerVoid() {
```

Comment:
I suppose such a trivial lowering is not needed as well, since it can be handled by a simple check in `ClassCodegen.generateMethod`

Reply:
@udalov I've just had a look and it's a bit more complicated since we also need to ensure that no other lowering touches the constructor of an annotation class.

Specifically, default argument stub generation needs a special case to avoid generating a stub for the constructor (which also replaces the default arguments with IrErrorExpressions) and enum class lowering needs a special case to avoid lowering the GetEnumValue constructs. Right now this only works since the IrConstructor is no longer being visited as it is only accessible through the initializers of corresponding properties. This is then how AnnotationCodegen gets the relevant information for default values in the backend.

I can certainly remove AnnotationLowering and add these three special cases, but in general this is a bit fragile since it depends on parts of the ir not being lowered. At least for the jvm ir backend I would rather go the other direction and replace annotation classes in the ir with something more specialized which is not a subtype of IrClass. What do you think?

Reply:
You're right, it does make sense to leave this lowering right now. But what do you mean specifically by
> something more specialized which is not a subtype of IrClass

?

Reply:
I mean having special constructs in the ir for annotation classes and instantiations.

Right now annotations are just special IrClasses with lots of implicit restrictions. This does allow for some code sharing, but adds a number of special cases. For example, annotation constructor calls have completely different codegen, default arguments for annotations are completely different from default arguments for normal constructors, annotation fields don't really have to be lowered and have very restricted types, etc.

In the end we don't really want to run all that many lowerings on annotation classes and we have to add special cases for them during codegen anyway. This begs the question if they shouldn't be represented with a different structure in the first place.

Reply:
We don't have a platform-specific IR at the moment, and while your points are totally valid, I'm not sure if annotations alone are a good reason to diverge the IR in this way. I agree that annotation classes are a special case most of the time though.

Maybe we could rather have this lowering transform annotation classes to interfaces, since they are represented as interfaces on JVM, storing the fact that the original class was an annotation in `IrDeclaration.origin` (for the codegen to generate ACC_ANNOTATION)? We'll still need to find a way to store default argument values somewhere (a mutable map in `JvmBackendContext` would be the first approximation), but at least we'll keep using the common IR and all lowerings will behave more consistently seeing annotation classes as interfaces rather than normal classes.

In any case, even if this works, it's not critical and can be postponed. For now, we can simply remove annotation constructors in this lowering.

Reply:
Are annotation classes less special in the other backends? :)

Just to be clear though, I totally agree that this is not something we should focus on now. The current implementation works (or will work after this PR gets merged anyway) and it's not going to break with any new lowering that's using the existing IrElementVisitor apis.